Add 'Settings' menu for `upstream-patches-ui`

davidbtadokoro opened the pull request:
This PR has the purpose of adding a 'Settings' menu for the `upstream-patches-ui` feature. A 'Settings' menu refers to an interface inside the feature to all its settings, mostly located in the `lore.config` file.

The 'Settings' menu is located in the 'Dashboard' screen, and, inside it, there is a submenu for each setting. At the moment, only the 'lists' and 'download_path' settings are configurable via the feature.

It is worth noting that any future setting (be it inside `lore.config` or not) should have a dedicated sub-menu inside the 'Settings' menu.

Closes: #829 
Closes: #811

Commit:
etc: lore.config: Add documentation to the file

The lore.config doesn't contain any documentation.

This commit adds a documentation that briefly describes the file and all
its configurations.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

Commit:
src: upstream_patches_ui: Add 'Register/Unregister Mailing Lists' menu

The user has one chance to manage the registered mailing lists that is
when the feature is first launched. After that, the only way to
add/remove lists is by manually editing the `lore.config` file, which
is both impractical and error prone.

This commit adds a menu entry inside the 'Dashboard'->'Settings' named
'Register/Unresgister Mailing Lists' that allows the user to revisit the
screen mentioned above and set the desired mailing lists available in
lore.kernel.org.

Note that this commit also adds the menu 'Settings', as the addition or
removal of mailing lists is a setting and, when we need to add new menus
related to settings, this top menu is already in place.

Closes: #811

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

rodrigosiqueira commented on src/upstream_patches_ui.sh:
```diff
@@ -193,6 +200,35 @@ function show_bookmarked_series_details()
+    0) # OK
+      case "$menu_return_string" in
+        1) # Mailing Lists
+          screen_sequence['SHOW_SCREEN']='register'
```
If I follow: Setting > Mailing Lists, I'll only have `OK` and `EXIT` options which make sense for the first screen. If I choose the OK option, it will send me back to the dashboard, which is not intuitive. It is time to parameterize the `register` option to enable it to return to the previous menu.

rodrigosiqueira commented on src/upstream_patches_ui.sh:
```diff
@@ -193,6 +200,35 @@ function show_bookmarked_series_details()
+  local message_box
+  local ret
+
+  menu_list_string_array=('Mailing Lists')
```
I think the name `Mailing List` is not good since it does not give the action idea. Maybe we can try something like `Follow Mailing List` or similar.

rodrigosiqueira commented on src/lib/dialog_ui.sh:
```diff
@@ -178,7 +178,7 @@ function create_simple_checklist()
 
   for item in "${_menu_list_string_array[@]}"; do
-    if [[ -n "${_check_statuses}" && -n ${_check_statuses["$index"]} ]]; then
+    if [[ "${_check_statuses["$index"]}" == 1 ]]; then
```
Can we have a test that exposes this bug?

rodrigosiqueira commented on src/upstream_patches_ui.sh:
```diff
@@ -402,31 +401,46 @@ function list_patches()
   fi
 
-  retrieve_available_mailing_lists "$flag"
+  create_loading_screen_notification "Retrieving available mailing lists from lore.kernel.org"
```
Single-quotes.

rodrigosiqueira commented on src/upstream_patches_ui.sh:
```diff
@@ -402,31 +401,46 @@ function list_patches()
+  # Put check marks on mailing lists already registered
+  for list in "${menu_list_string_array[@]}"; do
+    check_statuses["$index"]=0
+    # TODO: This may be inefficient, but we can't use partial matches as lists that are
```
I guess this is not a big deal since the number of lists is very small.

rodrigosiqueira commented on src/upstream_patches_ui.sh:
```diff
@@ -443,6 +457,9 @@ function register_mailing_list()
       handle_exit "$ret"
       ;;
+    3) # Return
+      screen_sequence['SHOW_SCREEN']='dashboard'
```
I guess that my comment in the second commit makes sense here.

rodrigosiqueira commented on src/lib/dialog_ui.sh:
```diff
@@ -295,6 +295,81 @@ function create_message_box()
+  flag=${flag:-'SILENT'}
+
+  box_title='Directory Selection Help'
+  message_box='There are 3 regions in the Directory Selection screen:'$'\n'
+  message_box+='- [Upper Box]: list of directories in the current path.'$'\n'
+  message_box+='- [Lower Box]: the current path (input box).'$'\n'
+  message_box+='- [Buttons]: "OK" to confirm directory path, "Cancel" to cancel selection and "Help" to show this screen.'$'\n'$'\n'
+  message_box+='To move between the regions, use the <TAB> key.'$'\n'
+  message_box+='To complete the current path with the highlighted directory, use the <SPACE> key.'$'\n'
+  message_box+='Typing while in the Upper Box or the Lower Box alters the current path.'
```
We have a better way to deal with this :) Check `load_module_text`.

rodrigosiqueira commented:
@melissawen

Commit:
src: upstream_patches_ui: Add menus for kernel tree settings

We need settings for defining a Linux kernel tree and a target branch of
this tree to perform tasks like applying patches. In other words, when
we implement actions like applying patches inside the feature, we will
need to define those two settings to determine which branch from which
tree to apply those patches.

This commit does this by adding two menus inside 'Dashboard'->'Settings'
named 'Kernel Tree Path' and 'Kernel Tree Target Branch' for setting a
Linux kernel tree and a target branch, respectively.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

davidbtadokoro commented on src/upstream_patches_ui.sh:
```diff
@@ -202,11 +202,17 @@ function show_bookmarked_series_details()
 # Screen that shows all types of settings available.
 function show_settings_screen()
 {
+  local lore_config_path="${PWD}/.kw/lore.config"
```
Hi, @rodrigosiqueira it doesn't create. Already made an issue for it, but I think is time to make a PR.
Thanks for the reminder!

davidbtadokoro commented on src/lib/dialog_ui.sh:
```diff
@@ -365,6 +365,62 @@ function create_directory_selection_help_screen()
+  cmd+=" '${height}' '${width}' '0'"
+  # Add choices (${choices["${choice}"]} refers to the choice's description)
+  for choice in "${!_choices[@]}"; do
+    cmd+=" $'${choice}' $'${_choices["${choice}"]}' off"
```
You are right. This change request slipped through in my last force-push. Going to work on it ASAP.

rodrigosiqueira commented on tests/lib/lore_test.sh:
```diff
@@ -421,4 +421,22 @@ function test_save_new_lore_config()
+  }
+  export -f git
+
+  get_kernel_tree_branches "$kernel_tree_path" 'branches'
```
Where this `$kernel_tree_branches` came from?
Also, I think that a test that creates a fake repository with multiple branches is more valuable for this case; in other words, I recommend you not use the `git` mock and use the function in a fake repository created by the test. I think we have some examples of that in the kw code.

davidbtadokoro replied:
Hi @rodrigosiqueira 
That was my first approach as I agree that mocking the `git` in this superficial way is not ideal. The tests worked perfectly but I was met with this message when trying to the sample repository of the project:

<img width="977" alt="Screen Shot 2023-06-10 at 20 18 41" src="https://github.com/kworkflow/kworkflow/assets/95055784/b537f289-d107-4625-ba08-247b9f919913">

My approach was to create a sample git repository and copy it to the project's tree. From what I understand this warning means that a submodule (a git repo insider another) must be declared together with a remote. Is it okay to ignore this message or am I making it wrong?

rodrigosiqueira replied:
In the setup phase, you can create a temporary folder followed by `git init`; that way, you will have a valid git repository in a temporary folder. In the tear_down phase, you can delete the temporary folder.

davidbtadokoro replied:
Thanks a lot! I'm sure that this approach will do 😄

Commit:
src: upstream_patches_ui: Add 'Save Patches To' menu

The `upstream-patches-ui` feature relies on a configuration named
'download_to' stored in `etc/lore.config` that defines the directory
where patch files will be stored. The only way to set it is to manually
alter the file, which is not ideal as the user may input an invalid path
and needs to know beforehand where it is defined.

This commit adds a menu entry named 'Save Patches To' located inside
'Dashboard'->'Settings' that allows the user to set the configuration,
now renamed to 'save_patches_to' for accuracy, through the feature's
interface.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

Commit:
src: lib: dialog_ui: Fix `create_simple_checklist` checkmark bug

The function defined in `dialog_ui.sh` that creates a dialog simple
checklist has the capability of marking the already checked entries of
the list. However, that capability has a bug that results in a false
positive (i.e., mark unchecked entries). The bug is due to the function
testing if the `check_statuses` (array used to determine the checked
entries) positions are set, not if they are set to 1 (the correct
approach). This resulted in positions set with anything, including 0
(which should mean unchecked), to be checked.

This commit fixes this bug by making the appropriate test. It also
updates the test suite to replicate the conditions of the bug, i.e., set
a position in the `check_statuses` with 0 and assert that this entry is
unchecked.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

Commit:
src: upstream_patches_ui: Revise 'Register/Unregister Mailing Lists' screen

The screen used to add/remove mailing lists and the function responsible
for displaying it, have some issues:
1) Lists already registered don't appear with checkmarks.
2) There is no loading screen while waiting for the fetch of available
   lists from lore.kernel.org.
3) The screen's title 'Lore List' is not precise and somewhat confusing.
4) The screen's message box is "It looks like that you don't have any
   lore list registered. Please, select one or more of the list below:"
   no matter if there are indeed lists registered.
5) The function's name `register_mailing_list` doesn't follow the
   convention of the 'show' prefix and may be confused with the
   `registered_mailing_list` function (also defined in the file).
6) The key used to identify the screen sequence is 'register' which can
   be enhanced.

This commit revises the screen and the function to address all the
issues above. The issues were handled by:
1) Checking already registered lists.
2) Adding a loading screen.
3) Changing title to 'Mailing Lists'.
4) Only showing the message box when there are no lists registered.
5) Change the function name to `show_mailing_lists_screen`.
6) Change the key to 'mailing_lists'.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

rodrigosiqueira commented:
Thanks for all the improvements in this PR, merged into the unstable branch.
